### PR TITLE
Rework ScoreProcessor score calculation methods to fix various issues

### PR DESCRIPTION
**Pull request by smoogipoo**
# Purpose

Forget everything you know about the methods `ScoreProcessor` exposes to calculate score. It now exposes a few different methods for computing total scores:
- **`ComputeFinalScore(ScoringMode mode, ScoreInfo scoreInfo)`**
    - Does not require `ApplyBeatmap` to be called.
    - To be used when the score is known to be complete (all of its statistics are populated).
- **`ComputePartialScore(ScoringMode mode, ScoreInfo scoreInfo)`**
    - **Requires** `ApplyBeatmap` to be called. This means you need to go through the full beatmap conversion process to use this method.
    -  To be used when it is not known whether the score is complete (such as multiplayer leaderboards).
- **`ComputeFinalLegacyScore(ScoringMode mode, ScoreInfo scoreInfo, int maxAchievableCombo)`**
    - Does not require `ApplyBeatmap` to be called.
    - This is a catch-all method for computing the final score when a score may come from a legacy source requiring a few hacks:
        - For legacy scores, the beatmap's max combo comes from an external source (e.g. diffcalc or API source) rather than statistics. For lazer scores, the beatmap's max combo can be determined via the statistics directly.
        - For legacy mania scores, the accuracy comes from the statistics since full-GREAT scores give 100% accuracy just as full-PERFECT scores do in osu-stable. For lazer mania scores, the accuracy comes from the provided `ScoreInfo`.
- **`ComputeScore(ScoringMode mode, double accuracyRatio, double comboRatio, double bonusScore, int totalBasicHitObjects)`**
    - Does not require `ApplyBeatmap` to be called.
    - This method does the bulk of the actual score calculation, but is also exposed as a catch-all method for anything else not covered by the above.
    - I have no immediate uses for this in mind so it could be privated upon request.

Several issues have come up in the recent while that necessitated the above changes. I'll go through each one:

## Cubic behaviour of classic scoring

Brought up in https://github.com/ppy/osu/issues/17011#issuecomment-1054704041.

Classic scoring exhibited cubic-scaling behaviour since it used the rolling number of hitobjects to multiply into the score, meaning that only the final score value was valid and any values before that were drastically scaled down. This was recently rediscovered in a [Discord conversation](https://discord.com/channels/188630481301012481/188630652340404224/950646354319908884).

This is fixed by counting the total number of hitobjects during initial "autoplay" processing.

## Requirement for a method to compute total score

Initially brought up in https://github.com/ppy/osu-server-spectator/pull/105/files/a78d5fdb5d7eb352d6b544e207ef287c1f42b237#r816449394.

The new `ComputeFinalScore()` method is an immediate solution to this. I also foresee maybe wanting to reclaculate scores database-side, for which`ComputeFinalLegacyScore()` can be used (or viewed as reference implementation for osu-web).

## Incorrect use of the previous `GetScore()` methods

The previous `GetScore()` methods were wishy-washy with respect to whether they needed to have `ApplyBeatmap()` called on them or not. I noticed in passing that one method requiring `ApplyBeatmap()` was used without having done so:
https://github.com/ppy/osu/blob/a352a140bc145f342fadc300797a2b05ade3f920/osu.Game/Rulesets/Difficulty/PerformanceBreakdownCalculator.cs#L65-L68
The effect of this is not noticeable, but this looks very dangerous and susceptible to error.

# Testing:
- Solo spectator + osu! ruleset
- Solo spectator + osu! ruleset + classic mod
- Solo spectator + osu!mania ruleset
- Playlists + osu! ruleset
- Multiplayer VS in-game leaderboard
- Multiplayer spectator VS in-game leaderboard

### CODE REVIEW

**Comment by bdach** [comment]
I've pushed a small fixup commit for some unresolved inspections.

Skimmed through the diff and it looks quite good to me, I don't really have any objections to anything done here. Generally I would approve this if not for the fact that I haven't done any testing myself and would need more time to test. Either way a second opinion is most likely a good idea here.

**Comment by peppy** [comment]
@smoogipoo able to conflict resolve this one? you're more familiar with the code than I am.

**Review by peppy — APPROVED** [review]
Feels really good from a code perspective.

**Pull request merged**
